### PR TITLE
Setup CI/CD Pipeline for MSStats

### PR DESCRIPTION
**Pull request by anshuman-raina**
### **User description**
# Motivation and Context

Part of Dataset Benchmarking Task

## Changes

Please provide a detailed bullet point list of your changes.

- Added slurm file for job set up
- Added github actions
- Added benchmark script

## Testing

Please describe any unit tests you added or modified to verify your changes.

## Checklist Before Requesting a Review
- [ ] I have read the MSstats [contributing guidelines](https://github.com/Vitek-Lab/MSstatsConvert/blob/master/.github/CONTRIBUTING.md)
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have run the devtools::document() command after my changes and committed the added files

___

### **PR Type**
enhancement, tests

___

### **Description**
- Added a GitHub Actions workflow to automate running R scripts on HPC using Slurm, including SSH setup, file transfer, job submission, and output retrieval.
- Developed a comprehensive R benchmarking script leveraging MSstats and MSstatsConvert for data processing and analysis, with parallel execution for improved performance.
- Created a Slurm configuration script to define job parameters and execute the benchmarking script on HPC.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>slurm_poc.yml</strong><dd><code>Add GitHub Actions workflow for HPC integration using Slurm</code></dd></summary>
<hr>

.github/workflows/slurm_poc.yml

<li>Added a GitHub Actions workflow to run R scripts on HPC via Slurm.<br> <li> Configured SSH access and file transfer to HPC.<br> <li> Implemented job submission and monitoring using Slurm.<br> <li> Set up artifact upload for job output.<br>

</details>

  </td>
  <td><a href="https://github.com/Vitek-Lab/MSstats/pull/143/files#diff-16dcc86cb818520d01256299234c54091ccdc4babf9e1d4d3753ae0342dd100b">+48/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>benchmark.R</strong><dd><code>Add benchmarking script with data processing and analysis</code></dd></summary>
<hr>

benchmark/benchmark.R

<li>Introduced a benchmarking script using MSstats and MSstatsConvert.<br> <li> Implemented data processing and result calculation functions.<br> <li> Utilized parallel processing for efficiency.<br> <li> Calculated and displayed performance metrics.<br>

</details>

  </td>
  <td><a href="https://github.com/Vitek-Lab/MSstats/pull/143/files#diff-ebecf9b0d956a58fd4325183def13cb06be198780f79360d05f814c5de5bb179">+178/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config.slurm</strong><dd><code>Create Slurm configuration script for benchmarking job</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

benchmark/config.slurm

<li>Created a Slurm configuration script for job submission.<br> <li> Defined job parameters such as name, output, and resources.<br> <li> Configured the execution of the R benchmarking script.<br>

</details>

  </td>
  <td><a href="https://github.com/Vitek-Lab/MSstats/pull/143/files#diff-a3ec1e4b1091ecce5ec7f8f598f6c8bfb5f2c6d0bf60b57742b5f68d2aa68559">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    steps:
+    - name: Checkout Repository
+      uses: actions/checkout@v3
+
+    - name: Set Up SSH Access
+      run: |
+        mkdir -p ~/.ssh
+        echo "${{ secrets.SSH_PRIVATE_KEY }}" > ~/.ssh/id_rsa
+        chmod 600 ~/.ssh/id_rsa
```
Comment (tonywu1999) [inline .github/workflows/benchmark.yml]: I would follow ChatGPT's suggestion here.

File: benchmark/config.slurm [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+#SBATCH --job-name=msstats_benchmark_job
+#SBATCH --output=job_output.txt
+#SBATCH --error=job_error.txt
+#SBATCH --time=01:00:00         # Set the maximum run time
+#SBATCH --ntasks=1              # Number of tasks (one process)
+#SBATCH --cpus-per-task=8      # Use 8 CPU cores for the task
+#SBATCH --mem=128G              # Request 256GB of memory
```
Comment (tonywu1999) [inline benchmark/config.slurm]: Comment says 256, but says 128 here.

File: .github/workflows/benchmark.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        ssh-keyscan -H login-00.discovery.neu.edu >> ~/.ssh/known_hosts
```
Comment (tonywu1999) [inline .github/workflows/benchmark.yml]: I think we should follow chatGPT's suggestion here (the error handling one)

File: .github/workflows/benchmark.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          while squeue -j ${{ env.slurm_job_id }} | grep -q ${{ env.slurm_job_id }}; do
+            echo 'Job Id : ${{ env.slurm_job_id }} is still running...'
+            sleep 10
+          done
+          echo 'Job has completed.'
+        "
+
+    - name: Fetch Output
+      run: |
+        scp raina.ans@login-00.discovery.neu.edu:/home/raina.ans/R/job_output.txt job_output.txt
```
Comment (tonywu1999) [inline .github/workflows/benchmark.yml]: There seems to be a lot of dependency on using your `/home/raina.ans` folder.  Could we instead use a folder in the `/work/VitekLab` directory?  I think there's already a benchmarking folder in there.

File: benchmark/benchmark.R [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  
+  #calculate SD and mean
+  
+  
+  # Kept the code for Individual Boxplots
+  
+  # boxplot(human_comparisonResult$log2FC,
+  #         main = "Boxplot of log2FC for Human",
+  #         ylab = "log2FC",
+  #         col = "lightblue")
```
Comment (tonywu1999) [inline benchmark/benchmark.R]: If you're not using the code, could you remove the comments?

File: benchmark/benchmark.R [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+head(fragpipe_raw)
+
+fragpipe_raw$Condition = unlist(lapply(fragpipe_raw$Run, function(x){
+  paste(str_split(x, "\\_")[[1]][4:5], collapse="_")
+}))
+
+fragpipe_raw$BioReplicate = unlist(lapply(fragpipe_raw$Run, function(x){
+  paste(str_split(x, "\\_")[[1]][4:7], collapse="_")
+}))
```
Comment (tonywu1999) [inline benchmark/benchmark.R:140]: I thought the Fragpipe files already have BioReplicate and Condition information

Comment (tonywu1999) [reply benchmark/benchmark.R:140]: Could you use the dataset stored in the `/work/VitekLab/Data/MS/Benchmarking` directory?  It should have the name `MSstats.csv`

File: benchmark/benchmark.R [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    result = function() dataProcess(msstats_format, featureSubset = "topN", n_top_feature = 20, MBimpute = FALSE)
+  ),
+  list(
+    label = "Data process without Normalization",
+    result = function() dataProcess(msstats_format, normalization = "FALSE", n_top_feature = 20)
+  ),
+  list(
+    label = "Data process without Normalization with MBImpute False",
+    result = function() dataProcess(msstats_format, normalization = "FALSE", n_top_feature = 20, MBimpute = FALSE)
+  )
```
Comment (tonywu1999) [inline benchmark/benchmark.R]: Does `n_top_feature` parameter need to be initialized to anything here?  i thought it's only needed if featureSubset = "topN"

File: .github/workflows/benchmark.yml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        scp raina.ans@login-00.discovery.neu.edu:/home/raina.ans/R/job_error.txt job_error.txt
```
Comment (tonywu1999) [inline .github/workflows/benchmark.yml]: I'm wondering if it makes sense to use a login info for someone like Olga.  I'm not sure if she has an OOD account though.  Or maybe use my login.

Comment (tonywu1999) [reply .github/workflows/benchmark.yml]: How difficult would it be to use someone else's login.  What would we need to adjust?

File: benchmark/benchmark.R [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+# Define the tasks with descriptive labels
+data_process_tasks <- list(
+  list(
+    label = "Data process with Normalized Data",
+    result = function() dataProcess(msstats_format, featureSubset = "topN", n_top_feature = 20)
+  ),
+  list(
+    label = "Data process with Normalization and MBImpute False",
+    result = function() dataProcess(msstats_format, featureSubset = "topN", n_top_feature = 20, MBimpute = FALSE)
```
Comment (tonywu1999) [inline benchmark/benchmark.R:142]: Could you check if n_top_feature = 10 makes any difference?  20 may be a lot.

File: .github/workflows/benchmark.yml (hunk 1 above)
Comment (anshuman-raina) [reply .github/workflows/benchmark.yml]: Fixed

File: benchmark/config.slurm (hunk 2 above)
Comment (anshuman-raina) [reply benchmark/config.slurm]: Fixed.

File: .github/workflows/benchmark.yml (hunk 3 above)
Comment (anshuman-raina) [reply .github/workflows/benchmark.yml]: Fixed.

File: benchmark/benchmark.R (hunk 5 above)
Comment (anshuman-raina) [reply benchmark/benchmark.R]: Fixed.

**Pull request merged**